### PR TITLE
feat: added MT support for regulation worker

### PR DESCRIPTION
**Pull request by saurav-malani**
# Description

- This PR adds Multi-tenant mode support in regulation-worker.  
- Modified regulation-worker on demand destination detail to pub-sub model.

## Notion Ticket
https://www.notion.so/rudderstacks/regulation-worker-adaptations-for-multitenant-20ba9db9d45a443f9bc9f62def374413

### CODE REVIEW

File: config/backend-config/single_workspace.go [hunk 1]
```diff
@@ -35,6 +35,14 @@ func (wc *singleWorkspaceConfig) SetUp() error {
...
 		}
 		return nil
+	} else {
+		config, err := wc.getFromAPI(context.TODO())
+		if err != nil {
+			return err
+		}
+		for workspaceID := range config {
+			wc.workspaceID = workspaceID
+		}
```
Comment (atzoum) [inline config/backend-config/single_workspace.go:45]:
```suggestion
```

File: regulation-worker/cmd/main.go [hunk 2]
```diff
@@ -52,25 +54,41 @@ func main() {
...
 	}
-	dest := &destination.DestMiddleware{
+	dest := &destination.DestinationConfig{
 		Dest: backendconfig.DefaultBackendConfig,
 	}
-	workspaceId, err := dest.GetWorkspaceId(ctx)
+	// for multi-tenant mode set following:
+	// `"DEPLOYMENT_TYPE"="MULTITENANT"`
+	// "WORKSPACE_NAMESPACE"=<namespaceID>
+	// "HOSTED_SERVICE_SECRET"=<namespace secret>
```
Comment (atzoum) [inline regulation-worker/cmd/main.go:67]:
```suggestion
```

File: regulation-worker/cmd/main.go [hunk 3]
```diff
@@ -52,25 +54,41 @@ func main() {
 
 func Run(ctx context.Context) {
 	admin.Init()
+	backendconfig.Init()
 	if err := backendconfig.Setup(nil); err != nil {
 		panic(fmt.Errorf("error while setting up backend config: %v", err))
 	}
```
Comment (atzoum) [inline regulation-worker/cmd/main.go:60]:
```suggestion
	}
	backendconfig.DefaultBackendConfig.StartWithIDs(ctx, "")
	backendconfig.DefaultBackendConfig.WaitForConfig(ctx)
```

File: regulation-worker/internal/destination/destination.go [hunk 4]
```diff
@@ -2,98 +2,59 @@ package destination
...
-		}
-		return nil
-	}, boCtx); err != nil {
-		if bo.NextBackOff() == backoff.Stop {
-			pkgLogger.Debugf("reached retry limit...")
-			return map[string]backendconfig.ConfigT{}, err
-		}
+func (d *DestinationConfig) BackendConfigSubscriber(ctx context.Context) {
+	workspaceId := d.Dest.Identity().ID()
+	d.Dest.StartWithIDs(ctx, workspaceId)
```
Comment (atzoum) [inline regulation-worker/internal/destination/destination.go:52]:
```suggestion
```

File: regulation-worker/cmd/main.go [hunk 5]
```diff
@@ -52,25 +54,41 @@ func main() {
...
 	svc := service.JobSvc{
 		API: &client.JobAPI{
-			Client:         &http.Client{Timeout: config.GetDuration("HttpClient.regulationWorker.regulationManager.timeout", 60, time.Second)},
-			URLPrefix:      config.MustGetString("CONFIG_BACKEND_URL"),
-			WorkspaceToken: config.MustGetString("CONFIG_BACKEND_TOKEN"),
-			WorkspaceID:    workspaceId,
+			Client:    &http.Client{Timeout: config.GetDuration("HttpClient.regulationWorker.regulationManager.timeout", 60, time.Second)},
+			URLPrefix: config.MustGetString("CONFIG_BACKEND_URL"),
+			Token:     config.MustGetString("CONFIG_BACKEND_TOKEN"),
+			ID:        ID,
```
Comment (atzoum) [inline regulation-worker/cmd/main.go:90]: please use `identity.Identifier`

File: regulation-worker/internal/client/client.go [hunk 6]
```diff
@@ -8,40 +8,48 @@ import (
...
-	WorkspaceToken string
+	Client    *http.Client
+	ID        string
+	URLPrefix string
+	Token     string
+	Mode      deployment.Type
+}
+
+func (j *JobAPI) getURL() string {
+	if j.Mode == deployment.DedicatedType {
```
Comment (atzoum) [inline regulation-worker/internal/client/client.go:32]: minor: a switch statement might be more beautiful!

File: regulation-worker/internal/client/client.go [hunk 7]
```diff
@@ -8,40 +8,48 @@ import (
...
-	URLPrefix      string
-	WorkspaceToken string
+	Client    *http.Client
+	ID        string
+	URLPrefix string
+	Token     string
+	Mode      deployment.Type
+}
+
+func (j *JobAPI) getURL() string {
```
Comment (atzoum) [inline regulation-worker/internal/client/client.go:31]:
```suggestion
func (j *JobAPI) URL() string {
```

File: regulation-worker/internal/client/client.go [hunk 8]
```diff
@@ -8,40 +8,48 @@ import (
...
+	url := j.getURL()
 	pkgLogger.Debugf("making GET request to URL: %v", url)
 	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
 	if err != nil {
 		pkgLogger.Errorf("error while create new http request: %v", err)
 		return model.Job{}, err
 	}
 
-	req.SetBasicAuth(j.WorkspaceToken, "")
+	req.SetBasicAuth(j.Token, "")
```
Comment (atzoum) [inline regulation-worker/internal/client/client.go:52]: you can use `identity.Identifier` for authorization

File: regulation-worker/internal/client/client.go [hunk 9]
```diff
@@ -69,10 +76,10 @@ func (j *JobAPI) Get(ctx context.Context) (model.Job, error) {
 			return model.Job{}, fmt.Errorf("error while decoding job: %w", err)
 		}
 
-		userCountPerJob := stats.Default.NewTaggedStat("user_count_per_job", stats.CountType, stats.Tags{"jobId": jobSchema.JobID, "workspaceId": j.WorkspaceID})
+		userCountPerJob := stats.Default.NewTaggedStat("user_count_per_job", stats.CountType, stats.Tags{"jobId": jobSchema.JobID, "workspaceId": j.ID})
```
Comment (atzoum) [inline regulation-worker/internal/client/client.go:79]: it is not always a workspaceId

File: regulation-worker/internal/client/client.go [hunk 10]
```diff
@@ -69,10 +76,10 @@ func (j *JobAPI) Get(ctx context.Context) (model.Job, error) {
...
 		userCountPerJob.Count(len(jobSchema.UserAttributes))
 
-		job, err := mapPayloadToJob(jobSchema, j.WorkspaceID)
+		job, err := mapPayloadToJob(jobSchema, j.ID)
```
Comment (atzoum) [inline regulation-worker/internal/client/client.go:82]: you shouldn't need the `j.ID` here

**Review by atzoum — APPROVED** [review]

**Review by cisse21 — APPROVED** [review]

**Pull request merged**